### PR TITLE
Color filenames based on file type

### PR DESCRIPTION
**Pull request by jimhester**
So this addresses #14, but has at least a couple of drawbacks.

1. Colors are only output if tibble is loaded (could also be a positive depending on your outlook on things)
2. Because dplyr verbs to do dispatch on `[` methods, the attribute is not subset appropriately when using dplyr verbs and the colors are incorrect.

I think this feature is nice, but not sure if it is worth pursuing further if it is not going to work in all cases.

### CODE REVIEW

**Comment by hadley** [comment]
Can you give me a screenshot of `dir_info()`?

**Comment by jimhester** [comment]
One alternative implementation that I think would work with the dplyr verbs and purrr is to store the paths as a list column. I did not do this because of storage efficiency, but I guess it might be the only way forward.

**Comment by hadley** [comment]
Yeah, it might be better to leave this until dplyr/vctrs does better at preserving classes.

**Comment by jimhester** [comment]
https://github.com/r-lib/fs/pull/23/commits/91e64d27a1d8b31dd8e6aae4e4635bcf56fd3d8d uses list columns, which do work with dplyr. I think it is worth going this route for now and we can maybe switch the implementation in the future once there is better support.

**Comment by hadley** [comment]
That'll only work with `dir_info()`, right? So we lose the nice colouring for `dir_list()`, but that's probably a small price to pay

**Comment by jimhester** [comment]
We don't do a `stat` for `dir_list()`, so we don't have all the information we would need to color the names. We could potentially do the stat on demand in that case, or leave `dir_list()` as a simple character vector as it is currently.

Here is how it looks in RStudio with the material theme and the LS_COLORS envirionment variable set with the file from https://github.com/trapd00r/LS_COLORS/blob/master/LS_COLORS

screenshot 2017-12-28 11 53 52

**Comment by hadley** [comment]
Oooh, that's pretty nice.  

I wonder if we should provide standard colours for "R" extensions?

**Comment by jimhester** [comment]
I made a major change in how this works. Now we do not store any mode information with the path, we simply stat the paths on-demand when we are colorizing paths. This means we stat files twice for `file_info()` / `dir_info()`, but it also means we can use this coloring everywhere.

I also added a `path_tidy()` function, which outputs tidy paths and changed all the functions in **fs** to return them.
> Tidy paths always expand `~`, use `/` to delimit directories, never have multiple `/` or
trailing `/` and have colourised output based on the file type.

It is true this will result in more stat calls in general but they are cheap, and will only happen when printing to the console.

File: DESCRIPTION [hunk 1]
```diff
@@ -26,4 +26,8 @@ URL: https://github.com/r-lib/fs
 BugReports: https://github.com/r-lib/fs/issues
 Suggests: 
     testthat,
-    covr
+    covr,
+    pillar,
+    crayon
+Remotes:
```
Comment (hadley) [inline DESCRIPTION]: pillar is on CRAN, fyi

File: R/path.R [hunk 2]
```diff
@@ -33,13 +32,40 @@ path_norm <- function(path) {
...
   path <- enc2utf8(path)
 
-  enc2utf8(path.expand(path))
+  new_fs_filename(path.expand(path))
+}
+
+#' Tidy paths
+#'
+#' untidy paths are all different, tidy paths are all the same.
+#'
```
Comment (hadley) [inline R/path.R]: I think you can drop the new line here

File: R/path.R [hunk 3]
```diff
@@ -33,13 +32,40 @@ path_norm <- function(path) {
...
+  # convert `\\` to `/`
+  path <- gsub("\\\\", "/", path)
+
+  # convert multiple // to single /
+  path <- gsub("//+", "/", path)
+
+  # Remove trailing / from paths
+  path <- sub("/$", "", path)
+
+  path
```
Comment (hadley) [inline R/path.R]: I think it would be clearer to do `new_fs_filename` here?

File: R/path.R [hunk 4]
```diff
@@ -33,13 +32,40 @@ path_norm <- function(path) {
...
 }
 
 #' Provide the path to the users home directory
 #' @export
 path_home <- function() {
```
Comment (hadley) [inline R/path.R]: Not really germane to this PR, but I think we should pass `...` onto to `path()`

File: R/utils.R [hunk 5]
```diff
@@ -20,9 +20,15 @@ auto_name <- function(names) {
...
-
 is_windows <- function() {
   tolower(Sys.info()[["sysname"]]) == "windows"
 }
+
+# This is needed to avoid checking the class of fs_filename objects in the
+# tests.
+compare.fs_filename <- function(x, y) {
+  if (identical(class(y), "character")) {
+    class(x) <- "character"
```
Comment (hadley) [inline R/utils.R]: Setting the class of a base type is a slightly odd thing to do.

**Comment by hadley** [comment]
Will also close #20 and maybe #21.

**Comment by hadley** [comment]
And #2

File: R/utils.R (hunk 5 above)
Comment (jimhester) [reply R/utils.R]: Yeah you are right, I can just use `class(x) <- NULL` here and get the same result.

**Pull request merged**